vispy/visuals/tube.py

```diff
@@ -0,0 +1,171 @@
+        indices = []
+        for i in range(segments):
+            for j in range(tube_points):
+                ip = (i+1) % segments if closed else i+1
+                jp = (j+1) % tube_points
+
+                index_a = i*tube_points + j
+                index_b = ip*tube_points + j
+                index_c = ip*tube_points + jp
+                index_d = i*tube_points + jp
+
+                indices.append([index_a, index_b, index_d])
+                indices.append([index_b, index_c, index_d])
+
+        vertices = grid.reshape(grid.shape[0]*grid.shape[1], 3)
+
+        if colors is not None and vertex_colors is None:
+            vertex_colors = np.zeros(vertices.shape, dtype=np.float32)
+            vertex_colors[:, 0] = np.repeat(colors[:, 0], tube_points)
+            vertex_colors[:, 1] = np.repeat(colors[:, 1], tube_points)
+            vertex_colors[:, 2] = np.repeat(colors[:, 2], tube_points)
+
+        indices = np.array(indices, dtype=np.uint32)
+
+        MeshVisual.__init__(self, vertices, indices,
+                            vertex_colors=vertex_colors,
+                            face_colors=face_colors,
+                            color=color,
+                            shading=shading,
+                            mode=mode)
+                
+
+    def draw(self, transforms):
+        MeshVisual.draw(self, transforms)
+
+
+def frenet_frames(points, closed):
+    '''Calculates and returns the tangents, normals and binormals for
+    the tube.'''
+    tangents = np.zeros((len(points), 3))
+    normals = np.zeros((len(points), 3))
+
+    epsilon = 0.0001
+
+    # Compute tangent vectors for each segment
+    tangents = np.roll(points, -1, axis=0) - np.roll(points, 1, axis=0)
+    mags = np.sqrt(np.sum(tangents * tangents, axis=1))
+    tangents /= mags[:, np.newaxis]
+
+    # Get initial normal and binormal
+    t = np.abs(tangents[0])
+
+    smallest = np.argmin(t)
+    normal = np.zeros(3)
+    normal[smallest] = 1.
+
+    vec = np.cross(tangents[0], normal)
+
+    normals[0] = np.cross(tangents[0], vec)
+
+    # Compute normal and binormal vectors along the path
+    for i in range(1, len(points)):
+        normals[i] = normals[i-1]
+
+        vec = np.cross(tangents[i-1], tangents[i])
+        if norm(vec) > epsilon:
+            vec /= norm(vec)
+
+            theta = np.arccos(np.clip(tangents[i-1].dot(tangents[i]), -1, 1))
+            normals[i] = rotation_about_axis(vec, theta).dot(normals[i])
+
+    if closed:
+        theta = np.arccos(np.clip(normals[0].dot(normals[-1]), -1, 1))
+        theta /= len(points) - 1
+
+        if tangents[0].dot(np.cross(normals[0], normals[-1])) > 0:
+            theta *= -1.
+            
+        for i in range(1, len(points)):
+            normals[i] = rotation_about_axis(
+                tangents[i], theta*i).dot(normals[i])
+
+    binormals = np.cross(tangents, normals)
+
+    return tangents, normals, binormals
+                
+
+def rotation_about_axis(axis, angle):
```

Comment:
I think we already have something like this in `vispy.transforms`. Can you check? In any case, that's where this would belong if we don't have it.